clang/test/SemaSYCL/device_global.cpp

```diff
@@ -33,6 +33,65 @@ namespace {
 device_global<int> same_name; // OK
 }
 
+// TODO: add diagnostic for this case
+inline namespace other {
+device_global<int> same_name; // ILLEGAL: shadows "device_global" variable
+} // namespace other
+
+// TODO: add diagnostic for this case
+// inline namespace {
+//  namespace foo {               // ILLEGAL: namespace name shadows "::foo"
+//  }                             // namespace which contains "device_global"
+//                                // variable.
+//}
+
+struct BBar {
+private:
+  struct BarInsider {
+    static device_global<float> c;
+  };
+};
+
+struct ABar {
+  void method() {
+    // expected-error@+1{{'device_global' variables must be static or declared at namespace scope}}
+    static device_global<float> c;
+  }
+  struct BarInsider {
+    static device_global<float> c;
+    void method() {
+      // expected-error@+1{{'device_global' variables must be static or declared at namespace scope}}
+      static device_global<float> c;
+    }
+  };
+};
+
+template <typename T> void fooBar() {
+  static device_global<T> c;
+  device_global<T> d;
+}
+
+template <typename T> struct TS {
+private:
+  static device_global<T> a;
+  // expected-error@+1 {{'device_global' variables must be static or declared at namespace scope}}
+  device_global<T> b;
+  // FIXME: Why are both messages emitted
```

Comment:
Perhaps a missing nullptr return or missing invalidation of the field?

Reply:
That doesn't help. Still debugging the issue.

Reply:
Could be one error is generated when parsing un-instantiated template, followed by second error after instantiation.